src/crimson/osd/osd_operation_sequencer.h

```diff
@@ -1,80 +1,88 @@
-// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:nil -*-
 // vim: ts=8 sw=2 smarttab
 
 #pragma once
 
 #include <map>
-
+#include <seastar/core/condition-variable.hh>
 #include "crimson/common/operation.h"
 #include "osd/osd_types.h"
 
 namespace crimson::osd {
 
-template <typename>
-struct OperationComparator;
-
+// when PG interval changes, we are supposed to interrupt all in-flight ops.
+// but in the order in which the ops are interrupted are not determined
+// because they are scheduled by the seastar scheduler, if we just interrupt
+// them at seeing a different interval when moving to a new continuation. but
+// we are supposed to replay the ops from the same client targeting the same
+// PG in the exact order that they are received.
+//
+// the way how we address this problem is to set up a blocker which blocks an
+// op until the preceding op is unblocked if the blocked one is issued in a new
+// pg interval.
+//
+// here, the ops from the same client are grouped by PG, and then ordered by
+// their id which is monotonically increasing and unique on per PG basis, so we
+// can keep an op waiting in the case explained above.
 template <typename T>
-class OperationRepeatSequencer {
+class OpSequencer {
 public:
-  using OpRef = boost::intrusive_ptr<T>;
-  using ops_sequence_t = std::map<OpRef, seastar::promise<>>;
-
-  template <typename Func, typename HandleT, typename Result = std::invoke_result_t<Func>>
-  seastar::futurize_t<Result> start_op(
-      HandleT& handle,
-      epoch_t same_interval_since,
-      OpRef& op,
-      const spg_t& pgid,
-      Func&& func) {
-    auto& ops = pg_ops[pgid];
-    if (!op->pos) {
-      [[maybe_unused]] auto [it, inserted] = ops.emplace(op, seastar::promise<>());
-      assert(inserted);
-      op->pos = it;
+  template <typename HandleT, typename FuncT>
+  seastar::future<> start_op(HandleT& handle,
+                             epoch_t interval,
+                             uint64_t op_id,
+                             FuncT&& do_op) {
+    if (last_interval == 0) {
+      last_interval = interval;
     }
-
-    auto curr_op_pos = *(op->pos);
-    const bool first = (curr_op_pos == ops.begin());
-    auto prev_op_pos = first ? curr_op_pos : std::prev(curr_op_pos);
-    auto prev_ops_drained = seastar::now();
-    if (epoch_t prev_interval = prev_op_pos->first->interval_start_epoch;
-        !first && same_interval_since > prev_interval) {
-      // need to wait for previous operations,
-      // release the current pipepline stage
+    auto have_green_light = seastar::make_ready_future<>();
+    assert(op_id > last_op);
+    const uint64_t prev_op = std::exchange(last_op, op_id);
+    assert(interval >= last_interval);
+    if (interval > last_interval) {
```

Comment:
It looks to me there could be cases as follows:
1. the 1st op gets started in the current interval, and set last_interval to the current interval;
2. the 3rd op reaches here before the 2nd op, and gets started as last_interval is already the current interval

If so, then client requests ordered is broken.